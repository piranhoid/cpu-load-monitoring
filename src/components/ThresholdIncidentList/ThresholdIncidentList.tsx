import React from 'react';

import useSystemInfo from '../../contexts/systemInfo';

const ThresholdIncidentList = () => {
  const {
    tresholdIncidentList,
  } = useSystemInfo();

  return (
    <table className="table-auto w-full">
      <thead>
        <tr>
          <th>
            Status
          </th>
          <th>
            Last Value
          </th>
          <th>
            Date
          </th>
        </tr>
      </thead>
      <tbody>
        {tresholdIncidentList.map((thresholdIncident, index) => {
          return (
            <tr key={index}>
              <td className="text-center">
                {thresholdIncident.isExceedingLimit ? '🔥' : '👌🏻'}
              </td>
              <td className="text-center">
                {thresholdIncident.lastValue}
              </td>
              <td className="text-center">
                {thresholdIncident.date.toLocaleString()}
              </td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
};

export default ThresholdIncidentList;

