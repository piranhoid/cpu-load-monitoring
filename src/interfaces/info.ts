export interface LoadAverageInfo {
  date: Date;
  value: number;
}

export interface AverageListItem {
  index: string;
  value: number;
}

export interface ThresholdIncidentListItem {
  date: Date;
  lastValue: number;
  isExceedingLimit: boolean;
}
