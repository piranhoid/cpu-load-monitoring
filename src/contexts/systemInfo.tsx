import React, { createContext, useContext, useEffect, useState } from 'react';
import { ipcRenderer } from 'electron';
import { AverageListItem, LoadAverageInfo, ThresholdIncidentListItem } from '../interfaces/info';

type SystemInfoContextType = {
  loadAverage: number | undefined;
  averageList: AverageListItem[];
  isExceedingLimit: boolean;
  tresholdIncidentList: ThresholdIncidentListItem[];
};

type SystemInfoProviderProps = { children: React.ReactNode };

export const SystemInfoContext = createContext<Partial<SystemInfoContextType>>({
  loadAverage: undefined,
  averageList: [],
  isExceedingLimit: false,
  tresholdIncidentList: [],
});

export const SystemInfoProvider = ({ children }: SystemInfoProviderProps) => {
  const [loadAverage, setLoadAverage] = useState<number>();
  const [averageList, setAverageList] = useState<AverageListItem[]>([]);
  const [isExceedingLimit, setIsExceedingLimit] = useState(false);
  const [tresholdIncidentList, setTresholdIncidentList] = useState<
    ThresholdIncidentListItem[]
  >([]);

  useEffect(() => {
    ipcRenderer.on('getCpuInfoResponse', (_, arg: LoadAverageInfo) => {
      setLoadAverage(arg.value);
      setAverageList((prevState: AverageListItem[]) => {
        let values: AverageListItem[] = [...prevState];
        if (values.length === 10) {
          values = values.splice(1);
        }
        values.push({
          index: arg.date.toLocaleString([], { hour: '2-digit', minute: '2-digit', second: '2-digit' }),
          value: arg.value
        });
        return values;
      });
    });
    ipcRenderer.on('getExceedingLimitResponse', (_, arg) => {
      setIsExceedingLimit(arg.isExceedingLimit);
      setTresholdIncidentList((prevState: ThresholdIncidentListItem[]) => {
        const values: ThresholdIncidentListItem[] = [...prevState];
        values.push({
          lastValue: arg.lastValue,
          date: arg.date,
          isExceedingLimit: arg.isExceedingLimit,
        });
        return values;
      });
    });
    setInterval(() => {
      ipcRenderer.send('getExceedingLimit');
    }, 2000);
    ipcRenderer.send('getCpuInfo');
  }, []);

  return (
    <SystemInfoContext.Provider
      value={{
        loadAverage,
        averageList,
        isExceedingLimit,
        tresholdIncidentList,
      }}
    >
      {children}
    </SystemInfoContext.Provider>
  );
};

const useSystemInfo = () => useContext(SystemInfoContext);

export default useSystemInfo;
