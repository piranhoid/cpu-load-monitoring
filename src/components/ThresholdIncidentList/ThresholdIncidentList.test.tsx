import React from 'react';
import renderer from 'react-test-renderer';

import { ThresholdIncidentListItem } from '../../interfaces/info';
import { SystemInfoContext } from '../../contexts/systemInfo';

import ThresholdIncidentList from './ThresholdIncidentList';

describe('[components] ThresholdIncidentList', () => {
  const tresholdIncidentList: ThresholdIncidentListItem[] = [{
    lastValue: 5,
    date: new Date('2000-01-10 00:00:00'),
    isExceedingLimit: true
  }, {
    lastValue: 5,
    date: new Date('2000-01-10 00:00:00'),
    isExceedingLimit: true
  }, {
    lastValue: 5,
    date: new Date('2000-01-10 00:00:00'),
    isExceedingLimit: true
  }, {
    lastValue: 5,
    date: new Date('2000-01-10 00:00:00'),
    isExceedingLimit: false
  }];


  it('should render correctly', async () => {
    const component = renderer.create(
      <SystemInfoContext.Provider value={{ tresholdIncidentList }}>
        <ThresholdIncidentList />
      </SystemInfoContext.Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
