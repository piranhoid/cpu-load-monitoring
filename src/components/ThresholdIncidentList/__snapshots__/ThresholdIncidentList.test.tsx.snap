// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] ThresholdIncidentList should render correctly 1`] = `
<table
  className="table-auto w-full"
>
  <thead>
    <tr>
      <th>
        Status
      </th>
      <th>
        Last Value
      </th>
      <th>
        Date
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="text-center"
      >
        🔥
      </td>
      <td
        className="text-center"
      >
        5
      </td>
      <td
        className="text-center"
      >
        1/10/2000, 12:00:00 AM
      </td>
    </tr>
    <tr>
      <td
        className="text-center"
      >
        🔥
      </td>
      <td
        className="text-center"
      >
        5
      </td>
      <td
        className="text-center"
      >
        1/10/2000, 12:00:00 AM
      </td>
    </tr>
    <tr>
      <td
        className="text-center"
      >
        🔥
      </td>
      <td
        className="text-center"
      >
        5
      </td>
      <td
        className="text-center"
      >
        1/10/2000, 12:00:00 AM
      </td>
    </tr>
    <tr>
      <td
        className="text-center"
      >
        👌🏻
      </td>
      <td
        className="text-center"
      >
        5
      </td>
      <td
        className="text-center"
      >
        1/10/2000, 12:00:00 AM
      </td>
    </tr>
  </tbody>
</table>
`;
